exceptions.AssertionError: threadpool_manager

```
ERROR   1435578237.91 tunnel_community:1091  Got ping from {95.211.198.140:21004} (not responding)
ERROR   1435578237.95 tunnel_community:1091  Got ping from {145.94.62.234:7759} (not responding)
ERROR   1435578238.59             log:455   Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/pouwelse/GITHUB/tribler/Tribler/Core/Utilities/twisted_thread.py", line 55, in _reactor_runner
    reactor.run(installSignalHandlers=False)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1192, in run
    self.mainLoop()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1201, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 797, in runUntilCurrent
    f(*a, **kw)
  File "/home/pouwelse/GITHUB/tribler/Tribler/Core/APIImplementation/threadpoolmanager.py", line 29, in <lambda>
    reactor.callFromThread(lambda: self.register_task(task_name, self._reactor.callLater(delay, wrapper)))
  File "/home/pouwelse/GITHUB/tribler/Tribler/dispersy/taskmanager.py", line 36, in register_task
    assert not self.is_pending_task_active(name), name
exceptions.AssertionError: threadpool_manager 486

Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/pouwelse/GITHUB/tribler/Tribler/Core/Utilities/twisted_thread.py", line 55, in _reactor_runner
    reactor.run(installSignalHandlers=False)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1192, in run
    self.mainLoop()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1201, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 797, in runUntilCurrent
    f(*a, **kw)
  File "/home/pouwelse/GITHUB/tribler/Tribler/Core/APIImplementation/threadpoolmanager.py", line 29, in <lambda>
    reactor.callFromThread(lambda: self.register_task(task_name, self._reactor.callLater(delay, wrapper)))
  File "/home/pouwelse/GITHUB/tribler/Tribler/dispersy/taskmanager.py", line 36, in register_task
    assert not self.is_pending_task_active(name), name
exceptions.AssertionError: threadpool_manager 486
ERROR   1435578252.92 tunnel_community:1091  Got ping from {145.94.62.234:7759} (not responding)
ERROR   1435578252.98 tunnel_community:1091  Got ping from {95.211.198.140:21004} (not responding)
```

--- a/Tribler/Core/APIImplementation/threadpoolmanager.py
+++ b/Tribler/Core/APIImplementation/threadpoolmanager.py
@@ -19,25 +19,23 @@ class ThreadPoolManager(TaskManager):
         self._lock = RLock()
         self._logger = logging.getLogger(self.__class__.__name__)
 
-    def add_task(self, wrapper, delay=0, task_name=None):
-        assert wrapper
-
+    def _check_task_name(self, task_name):
         if not task_name:
             with self._lock:
                 self._auto_counter += 1
-            task_name = "threadpool_manager %d" % self._auto_counter
-        reactor.callFromThread(lambda: self.register_task(task_name, self._reactor.callLater(delay, wrapper)))
+                task_name = "threadpool_manager %d" % self._auto_counter
+        return task_name
+
+    def add_task(self, wrapper, delay=0, task_name=None):
+        assert wrapper
+
+        reactor.callFromThread(lambda: self.register_task(self._check_task_name(task_name), self._reactor.callLater(delay, wrapper)))
 
     def add_task_in_thread(self, wrapper, delay=0, task_name=None):
         assert wrapper
 
-        if not task_name:
-            with self._lock:
-                self._auto_counter += 1
-            task_name = "threadpool_manager %d" % self._auto_counter
-
         def delayed_call(delay, task_name):
-            self.register_task(task_name, self._reactor.callLater(delay, reactor.callInThread, wrapper))
+            self.register_task(self._check_task_name(task_name), self._reactor.callLater(delay, reactor.callInThread, wrapper))
 
         reactor.callFromThread(delayed_call, delay, task_name)
 

--- a/Tribler/Test/test_as_server.py
+++ b/Tribler/Test/test_as_server.py
@@ -100,9 +100,9 @@ class AbstractServer(BaseTestCase):
 
         delayed_calls = reactor.getDelayedCalls()
         if delayed_calls:
-            self._logger.debug("The reactor was dirty:")
+            self._logger.error("The reactor was dirty:")
             for dc in delayed_calls:
-                self._logger.debug(">     %s" % dc)
+                self._logger.error(">     %s" % dc)
         self.assertFalse(delayed_calls, "The reactor was dirty when tearing down the test")
         self.assertFalse(Session.has_instance(), 'A session instance is still present when tearing down the test')
 

--- a/Tribler/Test/test_bartercast4_community.py
+++ b/Tribler/Test/test_bartercast4_community.py
@@ -1,21 +1,29 @@
 # Written by Cor-Paul Bezemer
 # see LICENSE.txt for license information
 
-from Tribler.Test.test_as_server import TestAsServer
+from time import sleep
 
+from twisted.internet.task import LoopingCall
+
+from Tribler.Core.Utilities.twisted_thread import reactor
+from Tribler.Test.test_as_server import TestAsServer
 from Tribler.community.bartercast4.community import BarterCommunity, BarterCommunityCrawler
+from Tribler.community.bartercast4.statistics import BartercastStatisticTypes, _barter_statistics
 from Tribler.community.tunnel.tunnel_community import TunnelSettings
 from Tribler.dispersy.util import blocking_call_on_reactor_thread, call_on_reactor_thread
-from Tribler.community.bartercast4.statistics import BartercastStatisticTypes, _barter_statistics
-from Tribler.Core.Utilities.twisted_thread import reactor
-from time import sleep
+
 
 DEBUG = True
 PEER_STATEDIR = "data"
 
 
 class TestBarterCommunity(TestAsServer):
+
     """Start a tribler session and wait for the statistics to be increased."""
+
+    def __init__(self, *argv, **kwargs):
+        super(TestBarterCommunity, self).__init__(*argv, **kwargs)
+        self._test_condition_lc = None
 
     def test_local_stats(self):
         def do_stats_test():
@@ -33,34 +41,38 @@ class TestBarterCommunity(TestAsServer):
                     return
 
         self.startTest(do_stats_test)
-        pass
 
     def test_stats_messages(self):
-        is_finished = False
 
-        @call_on_reactor_thread
         def do_stats_messages():
-            # check that the crawler receives messages here
-            # tries_left = 60
-
-            rows = _barter_statistics.get_interactions(self.dispersy)
-            if len(rows) > 0:
-                assert True, "Some bartercast statistic was crawled"
-                is_finished = True
+            # check if the crawler receives messages
+            if len(_barter_statistics.get_interactions(self.dispersy)):
+                # Some bartercast statistic was crawled.
+                self._test_condition_lc.stop()
                 self.quit()
-                return
-            reactor.callLater(5, do_stats_messages)
 
-        def finished():
-            return is_finished
+        def has_finished():
+            return not self._test_condition_lc.running
 
         def noop():
-            assert True
+            pass
 
-        self.startTest(do_stats_messages)
-        self.CallConditional(300.0, finished, noop, u"Failed to crawl")
+        # Make it blocking so CallConditional can't do a check before the LoopingCall exists.
+        @blocking_call_on_reactor_thread
+        def start():
+            self._test_condition_lc = LoopingCall(do_stats_messages)
+            self._test_condition_lc.start(5, now=True)
 
-#    for future stuff; you can use this if you need another peer for some reason
+        @blocking_call_on_reactor_thread
+        def cleanup_and_fail(succeeded, err_msg):
+            self._test_condition_lc.stop()
+            self._test_condition_lc = None
+            self.assert_(succeeded, err_msg)
+
+        self.startTest(start)
+        self.CallConditional(300.0, has_finished, noop, u"Failed to crawl", assert_callback=cleanup_and_fail)
+
+    # for future stuff; you can use this if you need another peer for some reason
     def setupPeer(self):
         from Tribler.Core.Session import Session
 

